Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/HourPackageSyncTool.kt

```diff
@@ -0,0 +1,163 @@
+        invalidateCachedKeys(downloadConfig.invalidHourEtags)
+
+        val missingHours = availableLocations.mapNotNull {
+            determineMissingHours(it, forceSync)
+        }
+        if (missingHours.isEmpty()) {
+            Timber.tag(TAG).i("There were no missing hours.")
+            return SyncResult(successful = true, newPackages = emptyList())
+        }
+
+        Timber.tag(TAG).d("Downloading missing hours: %s", missingHours)
+        requireStorageSpace(missingHours)
+
+        val hourDownloads = launchDownloads(missingHours, downloadConfig)
+
+        Timber.tag(TAG).d("Waiting for %d missing hour downloads.", hourDownloads.size)
+        val downloadedHours = hourDownloads.awaitAll().filterNotNull().also {
+            Timber.tag(TAG).v("Downloaded keyfile: %s", it.joinToString("\n"))
+        }
+        Timber.tag(TAG).i("Download success: ${downloadedHours.size}/${hourDownloads.size}")
+
+        return SyncResult(
+            successful = hourDownloads.size == downloadedHours.size,
+            newPackages = downloadedHours
+        )
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal suspend fun launchDownloads(
+        missingHours: Collection<LocationHours>,
+        downloadConfig: KeyDownloadConfig
+    ): Collection<Deferred<CachedKey?>> {
+        val launcher: CoroutineScope.(LocationHours, LocalDate, LocalTime) -> Deferred<CachedKey?> =
+            { locationData, targetDay, targetHour ->
+                async {
+                    val cachedKey = keyCache.createCacheEntry(
+                        location = locationData.location,
+                        dayIdentifier = targetDay,
+                        hourIdentifier = targetHour,
+                        type = Type.LOCATION_HOUR
+                    )
+
+                    try {
+                        downloadTool.downloadKeyFile(cachedKey, downloadConfig)
+                    } catch (e: Exception) {
+                        // We can't throw otherwise it cancels the other downloads too (awaitAll)
+                        null
+                    }
+                }
+            }
+
+        val downloads = missingHours
+            .flatMap { location ->
+                location.hourData.map { Triple(location, it.key, it.value) }
+            }
+            .flatMap { (location, day, hours) ->
+                hours.map { Triple(location, day, it) }
+            }
+        Timber.tag(TAG).d("Launching %d downloads.", downloads.size)
+
+        return downloads.map { (location, day, missingHour) ->
+            withContext(context = dispatcherProvider.IO) {
+                launcher(location, day, missingHour)
+            }
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun expectNewHourPackages(cachedHours: List<CachedKey>, now: Instant): Boolean {
+        val previousHour = now.toLocalTime().minusHours(1)
```

Comment:
Is this UTC?

Reply:
> Is this UTC? It should be...

It is, I've renamed the method to make it more clear.

> And why is this line so different from expectNewDayPackages?

One calls `nowUTC` and the other get's it passed as parameter as to not call it two times and work on two different timestamps.

Reply:
> It is, I've renamed the method to make it more clear.

Kept the name for now, to prevent conflicts with other changes. Will change in 1.8.x.